service/src/main/java/uk/gov/hmcts/reform/fpl/service/CaseRoleService.java

```diff
@@ -0,0 +1,87 @@
+package uk.gov.hmcts.reform.fpl.service;
+
+import com.google.common.collect.Sets;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.scheduling.annotation.Async;
+import org.springframework.stereotype.Service;
+import uk.gov.hmcts.reform.authorisation.generators.AuthTokenGenerator;
+import uk.gov.hmcts.reform.ccd.client.CaseUserApi;
+import uk.gov.hmcts.reform.ccd.client.model.CaseUser;
+import uk.gov.hmcts.reform.fpl.config.SystemUpdateUserConfiguration;
+import uk.gov.hmcts.reform.fpl.enums.CaseRole;
+import uk.gov.hmcts.reform.fpl.exceptions.GrantCaseAccessException;
+import uk.gov.hmcts.reform.idam.client.IdamClient;
+
+import java.util.HashSet;
+import java.util.Set;
+
+import static java.util.stream.Collectors.toSet;
+
+@Slf4j
+@Service
+@RequiredArgsConstructor(onConstructor = @__(@Autowired))
+public class CaseRoleService {
+
+    private final IdamClient idam;
+    private final CaseUserApi caseUser;
+    private final AuthTokenGenerator authTokenGenerator;
+    private final SystemUpdateUserConfiguration userConfig;
+    private final OrganisationService organisationService;
+
+    public void grantAccessToUser(String caseId, String user, Set<CaseRole> roles) {
+        grantCaseAccess(caseId, Set.of(user), roles);
+    }
+
+    @Async
+    public void grantAccessToLocalAuthority(String caseId, String localAuthority, Set<CaseRole> roles,
+                                            Set<String> excludeUsers) {
+        Set<String> localAuthorityUsers = getUsers(caseId, localAuthority, excludeUsers, roles);
+        grantCaseAccess(caseId, localAuthorityUsers, roles);
+    }
+
+    private void grantCaseAccess(String caseId, Set<String> users, Set<CaseRole> roles) {
+        final Set<String> usersGrantedAccess = new HashSet<>();
+
+        try {
+            final String userToken = idam.authenticateUser(userConfig.getUserName(), userConfig.getPassword());
+            final String serviceToken = authTokenGenerator.generate();
```

Comment:
I guess minor, but we are logging twice now - for single user and then for multiple sharing. Would you believe there is benefit of doing this once? Maybe caching it quickly for couple minutes?

Reply:
i think these are separated method, which here are called from totally separated threads, i think from logical perspective it is fine to keep them like this. Dont want to pass tokens as method params. 
Cacheing can be considered at client level i think, service like this should not bother about cacheing, it is not its responsibility